Loot Table API

IotaBread opened the pull request:
A port of Fabric's Loot table API with a few name changes to match current Minecraft code

IotaBread committed:
Loot Table API

LambdAurora commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/LootEntryTypeRegistry.java:
```diff
@@ -0,0 +1,41 @@
+	/**
+	 * Registers a loot entry type serializer by its ID.
+	 *
+	 * @param id the loot entry's ID
```
```suggestion
	 * @param id the loot entry's identifier
```
From [CONTRIBUTING.md](https://github.com/QuiltMC/quilt-standard-libraries/blob/1.18/CONTRIBUTING.md):
> Please void the use of abbreviations in javadocs, except if those describe the name of a format, a library, etc.

LambdAurora commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/mixin/LootManagerMixin.java:
```diff
@@ -0,0 +1,59 @@
+
+	@Inject(method = "apply(Ljava/util/Map;Lnet/minecraft/resource/ResourceManager;Lnet/minecraft/util/profiler/Profiler;)V", at = @At("RETURN"))
+	private void apply(Map<Identifier, JsonElement> map, ResourceManager manager, Profiler profiler, CallbackInfo ci) {
+		Map<Identifier, LootTable> newTables = new HashMap<>();
```
```suggestion
		Map<Identifier, LootTable> newTables = new Object2ObjectOpenHashMap<>();
```

LambdAurora commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/event/LootTableLoadingCallback.java:
```diff
@@ -0,0 +1,48 @@
+			}
+	);
+
+	void onLootTableLoading(ResourceManager resourceManager, LootManager manager, Identifier id, QuiltLootTableBuilder table, LootTableSetter setter);
```
Might want to javadoc this.

Also, since you need to pass a lambda anyway, would it be worth it to use a proper context object or would this have negative effects that I don't see?

IotaBread replied:
Added a context object

LambdAurora commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/QuiltLootTableBuilder.java:
```diff
@@ -0,0 +1,107 @@
+
+import java.util.Collection;
+
+public class QuiltLootTableBuilder extends LootTable.Builder {
```
Javadoc would be nice

IotaBread replied:
Documented QuiltLootPoolBuilder and this

Leo40Git commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/LootEntryTypeRegistry.java:
```diff
@@ -0,0 +1,41 @@
+ * @see #register(Identifier, JsonSerializer)
+ */
+public interface LootEntryTypeRegistry {
+	LootEntryTypeRegistry INSTANCE = new LootEntryTypeRegistryImpl();
```
What's the stance on singletons vs. static classes again?

IotaBread committed:
Improvements

Add javadoc for QuiltLootPoolBuilder and QuiltLootTableBuilder, use a context object in LootTableLoadingCallback, fix checkstyle errors

LambdAurora commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/event/LootTableLoadingCallback.java:
```diff
@@ -0,0 +1,73 @@
+ */
+@FunctionalInterface
+public interface LootTableLoadingCallback {
+	class Context {
```
If we have a Context object, then maybe remove LootSetter and include it directly in Context?

IotaBread replied:
Done

IotaBread committed:
Move LootTableSetter to context object

Leo40Git commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/event/LootTableLoadingCallback.java:
```diff
@@ -0,0 +1,74 @@
+		public final LootManager manager;
+		public final Identifier id;
+		public final QuiltLootTableBuilder table;
+		public final LootTableSetter setter;
```
Why not just have a `setLootTable` method in this class that calls this setter?

IotaBread replied:
Done

IotaBread committed:
Add setLootTable method to context that calls the setter

IotaBread committed:
Fix build error

Leo40Git commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/event/LootTableLoadingCallback.java:
```diff
@@ -49,6 +49,15 @@ private Context(ResourceManager resourceManager, LootManager manager, Identifier
 		}
+
+		/**
+		 * Replace the loot table this context is for, using the same id.
```
```suggestion
		 * Replace the loot table this context is for, using the same identifier.
```

IotaBread replied:
Should I link to the field here?

Leo40Git replied:
It's probably unnecessary, considering the field is undocumented (and doesn't need documentation, IMHO).

IotaBread committed:
Update library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/event/LootTableLoadingCallback.java

Co-authored-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

Juuxel commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/LootEntryTypeRegistry.java:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @see #register(Identifier, JsonSerializer)
+ */
+public interface LootEntryTypeRegistry {
```
This isn't needed, modders can just use the vanilla registry: https://github.com/Juuxel/fabric/blob/8142ea6035410ca1ed547aef205e9e5c951b2104/fabric-loot-tables-v1/src/main/java/net/fabricmc/fabric/api/loot/v1/LootEntryTypeRegistry.java#L30

Juuxel commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/LootJsonParser.java:
```diff
@@ -0,0 +1,54 @@
+import net.minecraft.util.JsonHelper;
+import net.minecraft.util.Lazy;
+
+public final class LootJsonParser {
```
This API is also not needed, vanilla's `LootGsons` exposes this functionality. Additionally, the reflection code below is really cursed - I should've used accessors back then.

Juuxel commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/QuiltLootPool.java:
```diff
@@ -0,0 +1,48 @@
+ * @see	QuiltLootPoolBuilder
+ * @see #cast(LootPool)
+ */
+public interface QuiltLootPool {
```
Static accessor classes (with methods like `QuiltLootPools.getEntries(LootPool)`) might be nicer for modders to use, and don't require any sort of casting or implementing interfaces with mixin. (I opted for that approach in my largely forgotten loot table API v2 for Fabric API)

Juuxel commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/QuiltLootPoolBuilder.java:
```diff
@@ -0,0 +1,122 @@
+		return new QuiltLootPoolBuilder();
+	}
+
+	public static QuiltLootPoolBuilder of(LootPool pool) {
```
`copyOf` might be more descriptive

Juuxel commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/QuiltLootTable.java:
```diff
@@ -0,0 +1,45 @@
+ * @see	QuiltLootPoolBuilder
+ * @see #cast(LootTable)
+ */
+public interface QuiltLootTable {
```
Same here about the mixin-implemented interfaces vs static accessor classes if you decide to change them

Juuxel commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/QuiltLootTableBuilder.java:
```diff
@@ -0,0 +1,117 @@
+		return new QuiltLootTableBuilder();
+	}
+
+	public static QuiltLootTableBuilder of(LootTable table) {
```
`copyOf` might be more descriptive here too

Juuxel commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/QuiltLootPoolBuilder.java:
```diff
@@ -0,0 +1,122 @@
+
+		if (copyRolls) {
+			rolls(extendedPool.getRolls());
+		}
```
This is missing the `bonusRolls` of the original loot pool (like FAPI's loot table v1)

Juuxel commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/mixin/LootPoolMixin.java:
```diff
@@ -0,0 +1,68 @@
+
+	@Override
+	public List<LootPoolEntry> getEntries() {
+		return Arrays.asList(entries);
```
`ImmutableList.copyOf` is safer here. Otherwise, these lists are mutable but fixed-size, which is both unsafe (imo, loot tables should only be modified during data loading) and a bit useless because of the size restriction.

(Also applies to all other `List<X> getXs` methods here and in `LootTableMixin`)

Juuxel commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/QuiltLootTable.java:
```diff
@@ -0,0 +1,45 @@
+
+	List<LootFunction> getFunctions();
+
+	LootContextType getType();
```
There's no need to have this in the API, it's public in `LootTable` nowadays

Juuxel reviewed (changes requested):
Here's a hopefully helpful review. I see that this API has inherited some of my design mistakes from Fabric's loot API, so I encourage looking at the v2 API here for a much better design: https://github.com/Juuxel/fabric/tree/c72ff1311e686e412691d4344b99bf5677c44bcf

I've detailed the specifics in the comments, but just copy-pasting Fabric's loot API a) includes stuff that Mojang has since exposed and b) misses some stuff added later like `bonusRolls`.

IotaBread committed:
Update API according to Juuxel's suggestions

IotaBread commented:
Thanks for the suggestions, I have updated the pr according to them

Juuxel commented on library/data/loot_table/src/testmod/java/org/quiltmc/qsl/loot/test/LootTableTestMod.java:
```diff
@@ -0,0 +1,92 @@
+	@Override
+	public void onInitialize() {
+		// Test loot entry
+		Registry.register(Registry.LOOT_POOL_ENTRY_TYPE, new Identifier("quilt", "extended_tag"), new LootPoolEntryType(new TestSerializer()));
```
There's no need to test this anymore, I suppose, since there's no custom API use here :smile:

IotaBread replied:
I replaced the test with yours, is that fine?

Juuxel replied:
Yep, that's totally fine.

IotaBread committed:
Fix licenses

Juuxel commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/QuiltLootPoolBuilder.java:
```diff
@@ -0,0 +1,176 @@
+	 * @see QuiltLootPools#getBonusRolls(LootPool)
+	 */
+	public QuiltLootPoolBuilder bonusRolls(LootNumberProvider bonusRolls) {
+		access.setBonusRollsRange(bonusRolls);
```
This can be replaced with a super-calling `@Override` method (this was added to `LootPool.Builder` in 1.16)

Juuxel reviewed (changes requested):
Looks really nice now, just some minor comments

IotaBread committed:
Suggested changes, rewrite test

IotaBread committed:
Merge branch '1.18' into loot_table_api

IotaBread committed:
Fix build

Leo40Git commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/QuiltLootPoolBuilder.java:
```diff
@@ -49,6 +49,12 @@ public QuiltLootPoolBuilder rolls(LootNumberProvider rolls) {
 	}
 
+	@Override
+	public LootPool.Builder bonusRolls(LootNumberProvider bonusRolls) {
```
This should return `QuiltLootPoolBuiler`.

Leo40Git commented on library/data/loot_table/src/main/java/org/quiltmc/qsl/loot/api/QuiltLootPoolBuilder.java:
```diff
@@ -0,0 +1,170 @@
+	}
+
+	@Override
+	public LootPool.Builder bonusRolls(LootNumberProvider bonusRolls) {
```
```suggestion
	public QuiltLootPoolBuilder bonusRolls(LootNumberProvider bonusRolls) {
```

TheGlitch76 committed:
Use the correct return type for the bonusRolls method

Co-authored-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

TheGlitch76 commented:
Hi--this PR seems to be unmaintained right now. I'm converting it to a draft to signify that it's not in a mergable state.

Feel free to mark it as ready for review once you begin working on this again.

EnnuiL commented:
Due to inactivity and plus the fact that this port of Fabric Loot Tables API v1 would have been made obsolete by the new Fabric Loot API v2 (and plus, direct ports are to be discouraged unless there's a reason for them), I'm closing this PR so it can be succeeded by another one
